subsys/bluetooth/audio/ascs.c

```diff
@@ -285,198 +398,143 @@ void ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
-		const struct bt_bap_stream_ops *ops = stream->ops;
-
-		switch (state) {
-		case BT_BAP_EP_STATE_IDLE:
-			ep->receiver_ready = false;
-			bt_bap_stream_reset(stream);
-
-			if (ops->released != NULL) {
-				ops->released(stream);
-			}
-			struct bt_ascs_ase *ase = CONTAINER_OF(ep, struct bt_ascs_ase, ep);
+	switch (new_state) {
+	case BT_BAP_EP_STATE_IDLE:
+		ase_set_state_idle(ase);
+		break;
+	case BT_BAP_EP_STATE_CODEC_CONFIGURED:
+		ase_set_state_codec_configured(ase);
+		break;
+	case BT_BAP_EP_STATE_QOS_CONFIGURED:
+		ase_set_state_qos_configured(ase);
+		break;
+	case BT_BAP_EP_STATE_ENABLING:
+		ase_set_state_enabling(ase);
+		break;
+	case BT_BAP_EP_STATE_STREAMING:
+		ase_set_state_streaming(ase);
+		break;
+	case BT_BAP_EP_STATE_DISABLING:
+		ase_set_state_disabling(ase);
+		break;
+	case BT_BAP_EP_STATE_RELEASING:
+		ase_set_state_releasing(ase);
+		break;
+	default:
+		__ASSERT_PRINT("Invalid state %d", new_state);
+	}
+}
 
-			ase_free(ase);
+int ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
+{
+	struct bt_ascs_ase *ase = CONTAINER_OF(ep, struct bt_ascs_ase, ep);
+	const enum bt_bap_ep_state old_state = ascs_ep_get_state(&ase->ep);
+	int err;
 
-			break;
+	switch (state) {
+	case BT_BAP_EP_STATE_IDLE:
+		break;
+	case BT_BAP_EP_STATE_CODEC_CONFIGURED:
+		switch (old_state) {
+		case BT_BAP_EP_STATE_IDLE:
 		case BT_BAP_EP_STATE_CODEC_CONFIGURED:
-			switch (old_state) {
-			case BT_BAP_EP_STATE_IDLE:
-			case BT_BAP_EP_STATE_CODEC_CONFIGURED:
-			case BT_BAP_EP_STATE_QOS_CONFIGURED:
-			case BT_BAP_EP_STATE_RELEASING:
-				break;
-			default:
-				BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-					      bt_bap_ep_state_str(old_state),
-					      bt_bap_ep_state_str(ep->status.state));
-				return;
-			}
-
-			ep->receiver_ready = false;
-
-			if (ops->configured != NULL) {
-				ops->configured(stream, &ep->qos_pref);
-			}
-
-			break;
 		case BT_BAP_EP_STATE_QOS_CONFIGURED:
-			/* QoS configured have different allowed states
-			 * depending on the endpoint type
-			 */
-			if (ep->dir == BT_AUDIO_DIR_SOURCE) {
-				switch (old_state) {
-				case BT_BAP_EP_STATE_CODEC_CONFIGURED:
-				case BT_BAP_EP_STATE_QOS_CONFIGURED:
-				case BT_BAP_EP_STATE_DISABLING:
-					break;
-				default:
-					BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-						      bt_bap_ep_state_str(old_state),
-						      bt_bap_ep_state_str(ep->status.state));
-					return;
-				}
-			} else {
-				switch (old_state) {
-				case BT_BAP_EP_STATE_CODEC_CONFIGURED:
-				case BT_BAP_EP_STATE_QOS_CONFIGURED:
-				case BT_BAP_EP_STATE_ENABLING:
-				case BT_BAP_EP_STATE_STREAMING:
-					break;
-				default:
-					BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-						      bt_bap_ep_state_str(old_state),
-						      bt_bap_ep_state_str(ep->status.state));
-					return;
-				}
-			}
-
-			ep->receiver_ready = false;
+		case BT_BAP_EP_STATE_RELEASING:
+			break;
+		default:
+			goto invalid_transition;
```

Comment:
Could consider using a local `bool` as `bool valid_state_transition` and then have a `if (valid_state_transition)` at the end of the function to avoid disrupting the flow of the function.